File: src/generator/src/main/java/com/smvp/generator/scan/utils/ScanUtils.java
/*
 * Copyright (C) 2009 - 2012 SMVP.NET
 *
 * This is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation; either version 3 of
 * the License, or (at your option) any later version.
 *
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this software; if not, write to the Free
 * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
 */

package com.smvp.generator.scan.utils;

import com.google.gwt.core.ext.typeinfo.HasAnnotations;
import com.smvp.generator.scan.model.AnnotationScanModel;
import com.smvp.generator.scan.model.MethodScanModel;
import com.smvp.generator.utils.ClassUtils;

import java.lang.annotation.Annotation;
import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.List;

/**
 * The Class ScanUtils.
 *
 * @author Nguyen Duc Dung
 * @since 12/4/11, 11:45 AM
 */
public final class ScanUtils {

    private ScanUtils() {

    }

    public static List<AnnotationScanModel> getAnnotations(HasAnnotations object) {
        List<AnnotationScanModel> models = new ArrayList<AnnotationScanModel>();
        for (Annotation annotation : object.getAnnotations()) {
            AnnotationScanModel annotationScanModel = new AnnotationScanModel();
            annotationScanModel.setAnnotationClassName(annotation.annotationType().getName());
            for (Method method : annotation.annotationType().getMethods()) {
                if (method.getDeclaringClass() == annotation.annotationType()) {
                    MethodScanModel methodScanModel = new MethodScanModel();
                    methodScanModel.setName(method.getName());
                    methodScanModel.setReturnType(method.getReturnType().getName());
                    methodScanModel.setReturnValue(buildReturnValueString(ClassUtils.
                            invokeMethod(method, annotation)));
                    annotationScanModel.getMethods().add(methodScanModel);
                }
            }
            models.add(annotationScanModel);
        }
        return models;
    }

    public static String buildReturnValueString(Object returnValue) {
        if (returnValue instanceof Class) {
            Class clazz = (Class) returnValue;
            return clazz.getName() + ".class";
        } else if (returnValue instanceof String) {
            String st = (String) returnValue;
            return "\"" + st + "\"";
        }
        return returnValue.toString();
    }

}

File: src/generator/src/main/scala/com/smvp4g/generator/scan/utils/ScanUtils.scala
/*
 * Copyright (C) 2009 - 2012 SMVP4G.COM
 *
 * This is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation; either version 3 of
 * the License, or (at your option) any later version.
 *
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this software; if not, write to the Free
 * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
 */

package com.smvp4g.generator.scan.utils

import com.google.gwt.core.ext.typeinfo.HasAnnotations
import collection.mutable.ListBuffer
import com.smvp4g.generator.scan.model.{MethodScanModel, AnnotationScanModel}
import com.smvp4g.generator.utils.ClassUtils

/**
 * The Class ScanUtils.
 *
 * @author Nguyen Duc Dung
 * @since 5/23/12, 12:26 AM
 *
 */

object ScanUtils {

  def getAnnotations(target: HasAnnotations): List[AnnotationScanModel] = {
    val models = new ListBuffer[AnnotationScanModel]
    target.getAnnotations.foreach(annotation => {
      val annotationScanModel = new AnnotationScanModel()
      annotationScanModel.setAnnotationClassName(annotation.annotationType().getName)
      annotation.annotationType.getMethods.foreach(method => {
        if (method.getDeclaringClass == annotation.annotationType()) {
          val methodScanModel = new MethodScanModel
          methodScanModel.setName(method.getName)
          methodScanModel.setReturnType(method.getReturnType.getName)
          methodScanModel.setReturnValue(buildReturnValueString(ClassUtils.
            invokeMethod(method, annotation)))
          annotationScanModel.methods.add(methodScanModel)
        }
      })
      models += annotationScanModel
    })
    models.toList
  }

  def buildReturnValueString(returnValue: AnyRef): String = {
    if (returnValue.isInstanceOf[Class[_]]) {
      val clazz = returnValue.asInstanceOf[Class[_]]
      return clazz.getName + ".class"
    } else if (returnValue.isInstanceOf[String]) {
      val st = returnValue.asInstanceOf[String]
      return "\"" + st + "\""
    }
    returnValue.toString
  }

}
